### ISSUE TITLE
Allow webserving on an alternate port instead of port 80/443

### ISSUE THREAD

**Issue by rfay** [drud/ddev#481]
## What happened (or feature request):

In #126 we'll solve the problem of explaining gracefully to people that their port 80 is already occupied, but some people may want to leave it that way. Other people don't know how to find out what is taking up the port, etc., and may want to just move ddev's webserving to a different port. We should consider providing configuration for that. 

This type of change may be pretty intrusive, because we'll have to add the port number to every reference of the URL (when it's not 80/443). That's a bit of impact on the code, but it's also less useful for developers, because in general they would not want to see a port in their URL, they want to see URLs that are structurally equivalent to production.

However, we should consider providing it.

In addition, the closed #323 and #376  advocate for reporting to users the actual localhost port binding for the webserver, so that should be done in this issue.

**Comment by rickmanelius** [comment]
@rfay Not sure if this is also related to https://github.com/drud/ddev/issues/321.

**Comment by wmdbMattes** [comment]
Running different ports is mandatory for us

**Comment by psychomieze** [comment]
While using Windows with IIS changing the port would be really useful to use ddev, as reconfiguring IIS is quite annoying - changing the ports in the docker-compose.yaml manually works but is kind of a sucky workaround - so if you could make the port configurable, it would be really great.

**Comment by rfay** [comment]
I absolutely support this feature and appreciate the feedback. 

However, for IIS, does `net stop w3svc` not work these days any more? https://technet.microsoft.com/en-us/library/cc732317(v=ws.10).aspx

**Comment by psychomieze** [comment]
@rfay yes, stopping works (also re-configuring all IIS binding works) - but I want to keep my apps running in parallel (especially as I for example have .NET based REST services running via IIS which then communicate with the PHP based stuff). So yes, if I run one OR the other everything works, but running in parallel should work, too I think.

**Comment by rickmanelius** [comment]
Taking another stab at this to convert from the abstract into an active proposal and a decision. 

## Scoping

The first question to address is if we're properly scoping this. To be more explicit, is this something that is already solvable within what DDEV already provides? If so, this may be simply resolved through additional documentation. Alternatively (and more likely), there may be an engineering effort required.

In terms of understanding if we can already address this use case, it's worth reiterating that we already provide the ability to directly access the web container directly. For example, when I configure a "drupal7" project, I can either access it by going to one of the following:

* http://drupal.ddev.local
* https://drupal.ddev.local
* http://localhost:32771

Note that there are caveats to this approach.

1. The port assignment is random to avoid collisions. So http://localhost:32771 may become http://localhost:32783 or anything else.
2. The router is what currently houses the SSL termination, therefore you cannot get the encrypted version of the site by accessing the port directly.
3. Finding the direct port is not an easy task. We surface how to directly connect to the DB container within `ddev describe`, but one has to currently use `docker ps -a` to find the appropriate port.
4. From a context switching perspective, all development would be against http://localhost:port, which doesn't make it easy to togle between multiple projects.

To that end, I don't believe this would be seen as a workable solution because the changing port numbers would be tenable.

## Implications

@rfay already mentioned that there would be a few areas impacted by this change. Here's a non-exhaustive list:

1. Router: This would require changes from the default of accepting ports 80, 443, 8025, and 8036 and inserting an arbitrary quantity depending upon the values specified in ~/.ddev/config.yml and used within ~/.ddev/docker-compose.yml.
2. Config: We would need to add two new key => value pairs that would then serve as variables loaded in docker-compose.yml.
4. Collisions: I'm not entirely sure this would occur, but we might need to have additional error checking to confirm that two different projects specifying the same ports do not result in a conflict. This would be particularly important for developers spanning many projects on multiple teams.
5. Routing: We would need to ensure that we're properly connecting the router and containers. I don't see this as a major issue, but I'm a bit naive on the amount of effort.
6. Default vs Override Display: I would suggest that anyone using 80/443 as defaults not see the port number appended and only display the :port when an ovveride was present.
7. Documentation/Help Test: There are a few areas that need to be updated to reflect these nuances.

## Proposal

I can think of a few ways to solve this, but the one that feels the most viable is the following:

1. Expose two key => value pairs in ~/.ddev/config.yml
2. Covert the default docker-compose.yml template to use these values.
3. Ensure the regenerated ddev router adds these additional ports and maps them correctly.
4. When different than port 80/443, expose the port on ddev list and ddev describe.
5. Review documentation and helper text to reflect this.

Open for comments/feedback...

**Comment by cweagans** [comment]
I did some thinking on this today. I like the proposal above, but I think it's missing a point for the router to detect changes and restart + rebind to the right ports.

I wonder if we should do a `ddev set-option` and `ddev set-option --global` command. This way, we'd have a scriptable way to change values in `.ddev/config.yml` + `~/.ddev/config.yml`, and then because there's some command that would change the option, we can have the router restart happen at that point.

I'm envisioning something like `ddev set-option --global http.port 1234` or `ddev set-option php.version 5.6`.

**Comment by rfay** [comment]
The biggest problem here is that we don't currently have global configuration, which is a shame. We'll have to add that. 

Nothing changes the site's config, all changes result from changes in the ~/.ddev/router-compose.yml, which is generated every time a site starts up. We're changing ports on the router, not on any site. 

I like the `ddev set-option` idea. I could also live short-term with just adding the config and not providing a command to edit it.

**Comment by cweagans** [comment]
> Nothing changes the site's config, all changes result from changes in the ~/.ddev/router-compose.yml, which is generated every time a site starts up. We're changing ports on the router, not on any site.

Right. I was saying that a `ddev set-option` command could have other uses than just changing the router ports. For this task specifically, we'd only need to do the router part.

> I like the ddev set-option idea. I could also live short-term with just adding the config and not providing a command to edit it.

If we don't add it, we'll have to find another point when ddev can react to the change and kill and restart the router with a different port configuration. I'm fine with that, but I didn't see any other place where it would make sense for that work to happen.

**Comment by rfay** [comment]
Updating [Rick's proposal](https://github.com/drud/ddev/issues/481#issuecomment-358431652):

> 1. Expose two key => value pairs in ~/.ddev/config.yml 
The configuration change to allow the router to offer alternate ports is a global configuration change, and does not affect any site's config.yaml or docker-compose.yml.  We'll need to allow configuration of the 4 ports used by the router. Changes made will affect every site. **As a result, we need to introduce global configuration for the first time. I would assume we'll add a ~/.ddev/config.yaml. **

> 2. Convert the default docker-compose.yml template to use these values.

There is no change required to docker-compose.yml, it's not part of the equation here, as only the router is being changed.

> 3. Ensure the regenerated ddev router adds these additional ports and maps them correctly.

Every time a site changes, the ~/.router-compose.yaml is regenerated and the resulting router is rebuilt. We may need to add a command to kick this, but `ddev start` would normally just solve it. However, just changing ~/.ddev/config.yaml would not affect anything without a `ddev start` or something.

> 4. When different than port 80/443, expose the port on ddev list and ddev describe.

Good point, very important. I suspect this would already almost work, but it's important, especially because we probably don't have formatting for it.

> 5. Review documentation and helper text to reflect this.

* We'll have to update the existing docs that explain how to solve conflicts, 
* The error we give when a conflict is discovered needs to point to the docs. It probably already does.

**Comment by rfay** [comment]
Beginning to look at this, I find (I think) that the implementation of our HTTP_EXPOSE (for each container) capabilities assumes that you expose the same HTTP port on the router that is exposed on the container. Since we want to change only the port on the router (and certainly not on the container) this puts us in a bit of a bind.  HTTPS_EXPOSE is perhaps a bit more flexible, as it allows a mapping.

**Comment by rfay** [comment]
Impressively, the docker-compose.yml HTTP_EXPOSE capability we already have (per container) supports changing the port already.

In the docker-compose.yml on one project, you can edit the HTTP_EXPOSE like this and get the desired behavior:
HTTP_EXPOSE=8080:80,8025

So with that change on site1/.dev/docker-compose.yml you can move http to http://site1.ddev.local:8080 - but site2 stays at http://site2.ddev.local:80

Unfortunately, ddev doesn't currently harvest the port information, so ddev list and ddev describe don't know about the change. My next adventure is to take a look and see if they can.

Also, since this is (currently) a change made directly to docker-compose.yml (which over time we have lost enthusiasm for the user editing) we'd need to use an override technique, which isn't too hard. 

But given this, my earlier statement that all the work was with the router is incorrect. It seems to be all with the project. But you'd have to move the port of each project separately.

**Comment by rickmanelius** [comment]
I believe I'm all caught up on this. And I prefer to keep this as a project configuration, which I believe is where @rfay is trending towards based on that last comment, because being able to store in code and share among team members will ensure consistency. It will also make it more apparent to a user that is used to looking into config.yml or docker-compose.yml versus having to know/remember there is a global config option as well. To that end, I see a PR open, so I will move my testing/comments there to see how this is working.
